File: android/app/src/main/java/com/zevclip/sender/filetransfer/FileTransferReceiver.kt
package com.zevclip.sender.filetransfer

import java.io.File
import java.io.RandomAccessFile
import java.security.MessageDigest
import java.util.UUID
import kotlin.math.min

class FileTransferReceiver(
    private val rootDir: File
) {
    data class CompletedFile(
        val fileId: String,
        val file: File,
        val sha256: String
    )

    data class CompletionResult(
        val transferId: String,
        val destinationRoot: File,
        val files: List<CompletedFile>
    )

    private data class Session(
        val manifest: FileTransferManifest,
        val tempRoot: File,
        val destinationRoot: File,
        val resumeToken: String,
        val stateMachine: FileTransferStateMachine,
        val verifiedChunksByFileId: MutableMap<String, MutableSet<Long>>
    )

    private val sessionsByTransferId = mutableMapOf<String, Session>()

    @Synchronized
    fun accept(
        manifest: FileTransferManifest,
        requestedStreamCount: Int = manifest.requestedStreamCount,
        receiverMaximumStreams: Int = ZevLinkTransferProtocol.MAX_STREAM_COUNT
    ): FileTransferOfferResponse {
        require(!sessionsByTransferId.containsKey(manifest.transferId)) { "Transfer already exists." }
        require(rootDir.mkdirs() || rootDir.isDirectory) {
            "Could not create transfer staging directory ${rootDir.absolutePath}."
        }
        require(rootDir.usableSpace >= manifest.totalBytes) {
            "Not enough disk space. Need ${manifest.totalBytes} bytes; available ${rootDir.usableSpace} bytes."
        }

        val streamCount = ZevLinkTransferProtocol.negotiatedStreamCount(requestedStreamCount, receiverMaximumStreams)
        val resumeToken = UUID.randomUUID().toString().replace("-", "").lowercase()
        val destinationRoot = uniqueDestinationRoot(manifest.senderName.ifBlank { manifest.transferId })
        val tempRoot = File(File(rootDir, ".partial"), "${manifest.transferId}-$resumeToken")
        prepareStorage(manifest, tempRoot)

        sessionsByTransferId[manifest.transferId] = Session(
            manifest = manifest,
            tempRoot = tempRoot,
            destinationRoot = destinationRoot,
            resumeToken = resumeToken,
            stateMachine = FileTransferStateMachine(
                initialState = FileTransferState.ACCEPTED,
                totalBytes = manifest.totalBytes,
                selectedStreamCount = streamCount
            ),
            verifiedChunksByFileId = manifest.fileEntries.associate { entry ->
                entry.fileId to mutableSetOf<Long>()
            }.toMutableMap()
        )

        return FileTransferOfferResponse(
            transferId = manifest.transferId,
            state = FileTransferState.ACCEPTED,
            chunkSize = ZevLinkTransferProtocol.CHUNK_SIZE_BYTES,
            streamCount = streamCount,
            resumeToken = resumeToken,
            candidates = emptyList()
        )
    }

    @Synchronized
    fun verifiedRanges(transferId: String): List<TransferVerifiedFileRanges> {
        val session = session(transferId)
        return session.manifest.fileEntries.map { entry ->
            TransferVerifiedFileRanges(
                fileId = entry.fileId,
                verifiedRanges = compactRanges(session.verifiedChunksByFileId[entry.fileId].orEmpty())
            )
        }
    }

    @Synchronized
    fun writeChunk(
        transferId: String,
        fileId: String,
        chunkIndex: Long,
        data: ByteArray,
        chunkSha256: String? = null,
        activeStreamCount: Int = 1
    ) {
        val session = session(transferId)
        val entry = session.manifest.fileEntries.firstOrNull { it.fileId == fileId }
            ?: error("File not found.")
        if (session.verifiedChunksByFileId[fileId]?.contains(chunkIndex) == true) {
            return
        }
        val size = entry.size ?: error("Directory entries cannot receive chunks.")
        val range = ZevLinkTransferProtocol.byteRangeForChunk(size, chunkIndex)
        require(data.size.toLong() == range.length) { "Chunk size does not match byte range." }
        if (chunkSha256 != null) {
            require(sha256Hex(data) == chunkSha256) { "Chunk SHA-256 mismatch." }
        }

        positionedWrite(File(session.tempRoot, entry.relativePath), range.startOffset, data)

        val verifiedChunks = session.verifiedChunksByFileId.getValue(fileId)
        if (verifiedChunks.add(chunkIndex)) {
            if (session.stateMachine.state == FileTransferState.ACCEPTED ||
                session.stateMachine.state == FileTransferState.INTERRUPTED
            ) {
                session.stateMachine.markChunkRequestStarted(activeStreamCount)
            }
            session.stateMachine.markChunkVerified(range.length)
        }
    }

    @Synchronized
    fun complete(transferId: String): CompletionResult {
        val session = session(transferId)
        val completedFiles = session.manifest.fileEntries.map { entry ->
            val expectedChunks = ZevLinkTransferProtocol.chunkCount(entry.size ?: 0L)
            val verifiedChunks = session.verifiedChunksByFileId[entry.fileId].orEmpty()
            require(verifiedChunks.size.toLong() == expectedChunks) { "Transfer is missing chunks." }

            val tempFile = File(session.tempRoot, entry.relativePath)
            val actualSha256 = sha256Hex(tempFile)
            entry.sha256?.let { expected ->
                require(expected == actualSha256) { "File SHA-256 mismatch." }
            }
            CompletedFile(
                fileId = entry.fileId,
                file = File(session.destinationRoot, entry.relativePath),
                sha256 = actualSha256
            )
        }

        session.stateMachine.beginVerification()
        session.stateMachine.complete()
        session.destinationRoot.parentFile?.mkdirs()
        require(session.tempRoot.renameTo(session.destinationRoot)) {
            "Could not atomically move transfer into place."
        }
        sessionsByTransferId.remove(transferId)

        return CompletionResult(
            transferId = transferId,
            destinationRoot = session.destinationRoot,
            files = completedFiles
        )
    }

    @Synchronized
    fun cancel(transferId: String) {
        sessionsByTransferId.remove(transferId)?.tempRoot?.deleteRecursively()
    }

    private fun session(transferId: String): Session {
        return sessionsByTransferId[transferId] ?: error("Transfer was not found.")
    }

    private fun prepareStorage(manifest: FileTransferManifest, tempRoot: File) {
        tempRoot.mkdirs()
        manifest.entries.forEach { entry ->
            val file = File(tempRoot, entry.relativePath)
            when (entry.kind) {
                FileTransferEntryKind.DIRECTORY -> file.mkdirs()
                FileTransferEntryKind.FILE -> {
                    file.parentFile?.mkdirs()
                    RandomAccessFile(file, "rw").use { randomAccessFile ->
                        randomAccessFile.setLength(entry.size ?: 0L)
                    }
                }
            }
        }
    }

    private fun positionedWrite(file: File, offset: Long, data: ByteArray) {
        RandomAccessFile(file, "rw").use { randomAccessFile ->
            randomAccessFile.channel.use { channel ->
                var written = 0
                while (written < data.size) {
                    written += channel.write(java.nio.ByteBuffer.wrap(data, written, data.size - written), offset + written)
                }
            }
        }
    }

    private fun uniqueDestinationRoot(rawName: String): File {
        val base = rawName.replace(Regex("""[/\\:\u0000\r\n]+"""), " ").trim().ifBlank { "Transfer" }
        var candidate = File(rootDir, base)
        var suffix = 2
        while (candidate.exists()) {
            candidate = File(rootDir, "$base $suffix")
            suffix += 1
        }
        return candidate
    }

    private fun compactRanges(chunks: Set<Long>): List<TransferChunkRange> {
        if (chunks.isEmpty()) {
            return emptyList()
        }
        val sorted = chunks.sorted()
        val ranges = mutableListOf<TransferChunkRange>()
        var start = sorted.first()
        var previous = start
        sorted.drop(1).forEach { chunk ->
            if (chunk == previous + 1L) {
                previous = chunk
            } else {
                ranges += TransferChunkRange(start, previous + 1L)
                start = chunk
                previous = chunk
            }
        }
        ranges += TransferChunkRange(start, previous + 1L)
        return ranges
    }

    companion object {
        fun sha256Hex(data: ByteArray): String {
            return MessageDigest.getInstance("SHA-256")
                .digest(data)
                .joinToString("") { "%02x".format(it) }
        }

        fun sha256Hex(file: File): String {
            val digest = MessageDigest.getInstance("SHA-256")
            file.inputStream().use { input ->
                val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
                while (true) {
                    val read = input.read(buffer)
                    if (read <= 0) break
                    digest.update(buffer, 0, read)
                }
            }
            return digest.digest().joinToString("") { "%02x".format(it) }
        }
    }
}

File: ZevClip/FileTransfer/FileTransferReceiver.swift
import CryptoKit
import Foundation

enum FileTransferReceiverError: Error, LocalizedError {
    case transferAlreadyExists
    case transferNotFound
    case fileNotFound
    case invalidChunk
    case hashMismatch
    case missingChunks
    case ioFailure(String)

    var errorDescription: String? {
        switch self {
        case .transferAlreadyExists:
            return "Transfer already exists."
        case .transferNotFound:
            return "Transfer was not found."
        case .fileNotFound:
            return "Transfer file was not found."
        case .invalidChunk:
            return "Chunk range or size is invalid."
        case .hashMismatch:
            return "SHA-256 verification failed."
        case .missingChunks:
            return "Transfer is missing chunks."
        case .ioFailure(let message):
            return message
        }
    }
}

struct FileTransferCompletedFile: Equatable {
    let fileId: String
    let url: URL
    let sha256: String
}

struct FileTransferCompletionResult: Equatable {
    let transferId: String
    let destinationRoot: URL
    let files: [FileTransferCompletedFile]
}

final class FileTransferReceiver {
    private final class Session {
        let manifest: FileTransferManifest
        let tempRoot: URL
        let finalRoot: URL
        let resumeToken: String
        var stateMachine: FileTransferStateMachine
        var verifiedChunksByFileId: [String: Set<Int64>]

        init(
            manifest: FileTransferManifest,
            tempRoot: URL,
            finalRoot: URL,
            resumeToken: String,
            streamCount: Int
        ) throws {
            self.manifest = manifest
            self.tempRoot = tempRoot
            self.finalRoot = finalRoot
            self.resumeToken = resumeToken
            self.stateMachine = try FileTransferStateMachine(
                initialState: .accepted,
                totalBytes: manifest.totalBytes,
                selectedStreamCount: streamCount
            )
            self.verifiedChunksByFileId = Dictionary(
                uniqueKeysWithValues: manifest.fileEntries.map { ($0.fileId, Set<Int64>()) }
            )
        }
    }

    private let lock = NSLock()
    private let fileManager: FileManager
    private let downloadsRoot: URL
    private var sessionsByTransferId: [String: Session] = [:]

    init(
        downloadsRoot: URL = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first!
            .appendingPathComponent("ZevLink Transfers", isDirectory: true),
        fileManager: FileManager = .default
    ) {
        self.downloadsRoot = downloadsRoot
        self.fileManager = fileManager
    }

    func accept(
        manifest: FileTransferManifest,
        requestedStreamCount: Int? = nil,
        receiverMaximumStreams: Int = ZevLinkTransferProtocol.maxStreamCount
    ) throws -> FileTransferOfferResponse {
        let streamCount = try ZevLinkTransferProtocol.negotiatedStreamCount(
            requestedStreamCount: requestedStreamCount ?? manifest.requestedStreamCount,
            receiverMaximum: receiverMaximumStreams
        )
        let resumeToken = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
        let finalRoot = try uniqueDestinationRoot(for: manifest)
        let tempRoot = downloadsRoot
            .appendingPathComponent(".partial", isDirectory: true)
            .appendingPathComponent("\(manifest.transferId)-\(resumeToken)", isDirectory: true)

        lock.lock()
        defer { lock.unlock() }

        guard sessionsByTransferId[manifest.transferId] == nil else {
            throw FileTransferReceiverError.transferAlreadyExists
        }

        try prepareStorage(for: manifest, tempRoot: tempRoot)
        let session = try Session(
            manifest: manifest,
            tempRoot: tempRoot,
            finalRoot: finalRoot,
            resumeToken: resumeToken,
            streamCount: streamCount
        )
        sessionsByTransferId[manifest.transferId] = session

        return try FileTransferOfferResponse(
            transferId: manifest.transferId,
            state: .accepted,
            chunkSize: ZevLinkTransferProtocol.chunkSizeBytes,
            streamCount: streamCount,
            resumeToken: resumeToken,
            candidates: []
        )
    }

    func verifiedRanges(transferId: String) throws -> [TransferVerifiedFileRanges] {
        let session = try session(for: transferId)
        return try session.manifest.fileEntries.map { entry in
            try TransferVerifiedFileRanges(
                fileId: entry.fileId,
                verifiedRanges: Self.compactRanges(from: session.verifiedChunksByFileId[entry.fileId] ?? [])
            )
        }
    }

    func writeChunk(
        transferId: String,
        fileId: String,
        chunkIndex: Int64,
        data: Data,
        chunkSHA256: String? = nil,
        activeStreamCount: Int = 1
    ) throws {
        let session = try session(for: transferId)
        guard let entry = session.manifest.fileEntries.first(where: { $0.fileId == fileId }),
              let fileSize = entry.size else {
            throw FileTransferReceiverError.fileNotFound
        }
        if session.verifiedChunksByFileId[fileId]?.contains(chunkIndex) == true {
            return
        }

        let range = try ZevLinkTransferProtocol.byteRangeForChunk(fileSize: fileSize, chunkIndex: chunkIndex)
        guard Int64(data.count) == range.length else {
            throw FileTransferReceiverError.invalidChunk
        }
        if let chunkSHA256, Self.sha256Hex(data: data) != chunkSHA256 {
            throw FileTransferReceiverError.hashMismatch
        }

        try positionedWrite(data, to: fileURL(for: entry, root: session.tempRoot), offset: range.startOffset)

        lock.lock()
        defer { lock.unlock() }

        guard let lockedSession = sessionsByTransferId[transferId] else {
            throw FileTransferReceiverError.transferNotFound
        }
        if lockedSession.verifiedChunksByFileId[fileId, default: []].insert(chunkIndex).inserted {
            if lockedSession.stateMachine.state == .accepted || lockedSession.stateMachine.state == .interrupted {
                try lockedSession.stateMachine.markChunkRequestStarted(activeStreamCount: activeStreamCount)
            }
            try lockedSession.stateMachine.markChunkVerified(byteCount: range.length)
        }
    }

    func complete(transferId: String) throws -> FileTransferCompletionResult {
        let session = try session(for: transferId)
        var completedFiles: [FileTransferCompletedFile] = []

        for entry in session.manifest.fileEntries {
            guard let expectedSize = entry.size else {
                continue
            }
            let expectedChunks = try ZevLinkTransferProtocol.chunkCount(fileSize: expectedSize)
            let verifiedChunks = session.verifiedChunksByFileId[entry.fileId] ?? []
            guard verifiedChunks.count == expectedChunks else {
                throw FileTransferReceiverError.missingChunks
            }

            let tempURL = fileURL(for: entry, root: session.tempRoot)
            let actualSHA256 = try Self.sha256Hex(fileURL: tempURL)
            if let expectedSHA256 = entry.sha256, expectedSHA256 != actualSHA256 {
                throw FileTransferReceiverError.hashMismatch
            }
            completedFiles.append(
                FileTransferCompletedFile(
                    fileId: entry.fileId,
                    url: session.finalRoot.appendingPathComponent(entry.relativePath, isDirectory: false),
                    sha256: actualSHA256
                )
            )
        }

        try markVerifyingAndComplete(transferId: transferId)
        try fileManager.createDirectory(
            at: session.finalRoot.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try fileManager.moveItem(at: session.tempRoot, to: session.finalRoot)

        lock.lock()
        sessionsByTransferId.removeValue(forKey: transferId)
        lock.unlock()

        return FileTransferCompletionResult(
            transferId: transferId,
            destinationRoot: session.finalRoot,
            files: completedFiles
        )
    }

    func cancel(transferId: String) throws {
        let session = try session(for: transferId)
        lock.lock()
        sessionsByTransferId.removeValue(forKey: transferId)
        lock.unlock()
        try? fileManager.removeItem(at: session.tempRoot)
    }

    private func session(for transferId: String) throws -> Session {
        lock.lock()
        defer { lock.unlock() }
        guard let session = sessionsByTransferId[transferId] else {
            throw FileTransferReceiverError.transferNotFound
        }
        return session
    }

    private func prepareStorage(for manifest: FileTransferManifest, tempRoot: URL) throws {
        try fileManager.createDirectory(
            at: downloadsRoot,
            withIntermediateDirectories: true
        )

        if let available = try downloadsRoot
            .resourceValues(forKeys: [.volumeAvailableCapacityForImportantUsageKey])
            .volumeAvailableCapacityForImportantUsage,
           available < manifest.totalBytes {
            throw FileTransferReceiverError.ioFailure(
                "Not enough disk space. Need \(ByteCountFormatter.string(fromByteCount: manifest.totalBytes, countStyle: .file)); available \(ByteCountFormatter.string(fromByteCount: available, countStyle: .file))."
            )
        }

        try fileManager.createDirectory(at: tempRoot, withIntermediateDirectories: true)
        for entry in manifest.entries {
            let url = fileURL(for: entry, root: tempRoot)
            switch entry.kind {
            case .directory:
                try fileManager.createDirectory(at: url, withIntermediateDirectories: true)
            case .file:
                try fileManager.createDirectory(
                    at: url.deletingLastPathComponent(),
                    withIntermediateDirectories: true
                )
                let fileDescriptor = open(url.path, O_CREAT | O_TRUNC | O_RDWR, S_IRUSR | S_IWUSR)
                guard fileDescriptor >= 0 else {
                    throw FileTransferReceiverError.ioFailure("Could not create \(url.path): \(String(cString: strerror(errno)))")
                }
                defer { close(fileDescriptor) }
                if ftruncate(fileDescriptor, entry.size ?? 0) != 0 {
                    throw FileTransferReceiverError.ioFailure("Could not preallocate \(url.path): \(String(cString: strerror(errno)))")
                }
            }
        }
    }

    private func uniqueDestinationRoot(for manifest: FileTransferManifest) throws -> URL {
        let baseName = sanitizedName(manifest.senderName.isEmpty ? manifest.transferId : manifest.senderName)
        var candidate = downloadsRoot.appendingPathComponent(baseName, isDirectory: true)
        var suffix = 2
        while fileManager.fileExists(atPath: candidate.path) {
            candidate = downloadsRoot.appendingPathComponent("\(baseName) \(suffix)", isDirectory: true)
            suffix += 1
        }
        return candidate
    }

    private func sanitizedName(_ name: String) -> String {
        let illegal = CharacterSet(charactersIn: "/:\\\0").union(.newlines)
        let cleaned = name.components(separatedBy: illegal).joined(separator: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        return cleaned.isEmpty ? "Transfer" : cleaned
    }

    private func fileURL(for entry: FileTransferEntry, root: URL) -> URL {
        root.appendingPathComponent(entry.relativePath, isDirectory: entry.kind == .directory)
    }

    private func positionedWrite(_ data: Data, to url: URL, offset: Int64) throws {
        let fileDescriptor = open(url.path, O_WRONLY)
        guard fileDescriptor >= 0 else {
            throw FileTransferReceiverError.ioFailure("Could not open \(url.path): \(String(cString: strerror(errno)))")
        }
        defer { close(fileDescriptor) }

        try data.withUnsafeBytes { buffer in
            guard let baseAddress = buffer.baseAddress else {
                return
            }
            var written = 0
            while written < data.count {
                let result = pwrite(
                    fileDescriptor,
                    baseAddress.advanced(by: written),
                    data.count - written,
                    off_t(offset) + off_t(written)
                )
                guard result > 0 else {
                    throw FileTransferReceiverError.ioFailure("Could not write \(url.path): \(String(cString: strerror(errno)))")
                }
                written += result
            }
        }
    }

    private func markVerifyingAndComplete(transferId: String) throws {
        lock.lock()
        defer { lock.unlock() }
        guard let session = sessionsByTransferId[transferId] else {
            throw FileTransferReceiverError.transferNotFound
        }
        try session.stateMachine.beginVerification()
        try session.stateMachine.complete()
    }

    private static func compactRanges(from chunks: Set<Int64>) throws -> [TransferChunkRange] {
        let sorted = chunks.sorted()
        guard let first = sorted.first else {
            return []
        }

        var ranges: [TransferChunkRange] = []
        var start = first
        var previous = first
        for chunk in sorted.dropFirst() {
            if chunk == previous + 1 {
                previous = chunk
            } else {
                ranges.append(try TransferChunkRange(startChunk: start, endChunkExclusive: previous + 1))
                start = chunk
                previous = chunk
            }
        }
        ranges.append(try TransferChunkRange(startChunk: start, endChunkExclusive: previous + 1))
        return ranges
    }

    static func sha256Hex(data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    static func sha256Hex(fileURL: URL) throws -> String {
        guard let fileHandle = FileHandle(forReadingAtPath: fileURL.path) else {
            throw FileTransferReceiverError.ioFailure("Could not open \(fileURL.path) for hashing.")
        }
        defer { try? fileHandle.close() }

        var hasher = SHA256()
        while true {
            let data = try fileHandle.read(upToCount: 1024 * 1024) ?? Data()
            if data.isEmpty {
                break
            }
            hasher.update(data: data)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
}
